### PR TITLE
Financing Statement container

### PR DESCRIPTION
**Pull request by bryan-gilbert**
The form elements for a Financing statement. Includes edit and display mode.
Update to new names
Update to use default person constructor

### CODE REVIEW

File: ppr-ui/src/components/FormSectionHeader.vue [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import { createComponent } from '@vue/composition-api'
+
+export default createComponent({
+  props: {
+    label: String
+  },
+  setup() {
+    return {
+    }
+  }
```
Comment (WalterMoar) [inline ppr-ui/src/components/FormSectionHeader.vue]: Curious - this component looks fairly complete, so do we need the setup?

File: ppr-ui/src/financing-statement/FinancingStatement.vue [hunk 2]
```diff
@@ -0,0 +1,130 @@
+<template>
+  <v-card
+    class="mb-12"
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/FinancingStatement.vue]: ?

File: ppr-ui/src/financing-statement/FinancingStatement.vue [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+export default createComponent({
+  components: {
+    FormSectionHeader,
+    RegisteringParty
+  },
+  props: {
+    editing: {
+      default: false,
+      required: true,
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/FinancingStatement.vue]: This should be `required: false`, no? We don't have a `default` value on required properties.

File: ppr-ui/src/financing-statement/FinancingStatement.vue [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    value: {
+      required: true,
+      type: FinancingStatementModel
+    }
+  },
+
+  setup(props, { emit }) {
+    const items = ref<string[]>(FinancingStatementTypesCodeList)
+    const term = ref<number>(1)
+    const validTermDef = 'Term must be a number between 1 and 25'
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/FinancingStatement.vue]: Can `validTermDef` be removed and the string moved to `termRules`? I don't think it needs its own variable.

File: ppr-ui/src/financing-statement/FinancingStatement.vue [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      emit('input', new FinancingStatementModel(
+        props.value.type,
+        props.value.term,
+        newPerson // props.value.registeringParty
+      ))
+    }
+
+    // Callback function for emitting model changes made to the FS term
+    function updateTerm(newTerm: number): void {
+      if (props.value) {
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/FinancingStatement.vue]: With the `value` property being required, this check is no longer needed (make for better branch coverage).

File: ppr-ui/src/financing-statement/FinancingStatement.vue [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        emit('input', new FinancingStatementModel(
+          props.value.type,
+          newTerm, // props.value.term,
+          props.value.registeringParty))
+      }
+    }
+
+    // Callback function for emitting model changes made to the FS term
+    function updateType(newType: FinancingStatementTypes): void {
+      if (props.value) {
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/FinancingStatement.vue]: This check can also be removed.

File: ppr-ui/src/financing-statement/financing-statement-model.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import { FinancingStatementTypes } from '@/financing-statement/financing-statement-types'
+import { PersonModel } from '@/components/person-model'
+
+export default class FinancingStatementModel {
+
+  /*
+    Helper function to validate the term of a financing statement given a string value from an input field.
+  */
+  public static isValidTerm(value: string): string | boolean {
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/financing-statement-model.ts]: So static methods in the class before the member variables? Not opposed to that, just haven't seen it before.

File: ppr-ui/src/financing-statement/financing-statement-model.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import { FinancingStatementTypes } from '@/financing-statement/financing-statement-types'
+import { PersonModel } from '@/components/person-model'
+
+export default class FinancingStatementModel {
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/financing-statement-model.ts]: Perhaps a change for later, but I'm liking the idea of not having default exports, unless it's a special sort of helper utility.

https://basarat.gitbook.io/typescript/main-1/defaultisbad

File: ppr-ui/src/financing-statement/financing-statement-model.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    return isValid
+  }
+
+  private _type: FinancingStatementTypes
+  private _term: number
+  private _registeringParty: PersonModel
+
+  /**
+   *
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/financing-statement-model.ts:11]: Missing description for the constructor (which is of course boilerplate but might as well be consistent).

File: ppr-ui/src/router/routes.ts [hunk 10]
```diff
@@ -2,7 +2,7 @@ import About from '@/views/About.vue'
 import Home from '@/views/Home.vue'
 import SearchPage from '@/views/SearchPage.vue'
 import ResultsPage from '@/views/ResultsPage.vue'
-import FinancingStatement from '@/views/FinancingStatement.vue'
+import FinancingStatement from '@/views/FinancingStatementView.vue'
```
Comment (WalterMoar) [inline ppr-ui/src/router/routes.ts:5]: Do you want to import this as `FinancingStatementView`? Just asking, as long as we're consistent is what matters.

File: ppr-ui/src/views/FinancingStatementView.vue [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        <p>
+          Need copy here to instruct user as to how to create a financing statement.
+          This is a development only form and view. Future development will include; debtors, collateral, and secured parties
+        </p>
+        <section>
+          <p>{{ financingStatement }} </p>
+          <financing-statement
+            :editing="editing"
+            :value="financingStatement"
+            @input="updateModel"
```
Comment (WalterMoar) [inline ppr-ui/src/views/FinancingStatementView.vue]: This should be `v-model="financingStatement"`, which gets rid of `updateModel`.

File: ppr-ui/tests/unit/financing-statement/financing-statment.spec.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+      test('26', false)
+    })
+
+    it('term of empty is not valid', (): void => {
+      test('', false)
+    })
+
+    it('term of not a number input ("2b") is not valid', (): void => {
+      test('2b', false)
+    })
```
Comment (WalterMoar) [inline ppr-ui/tests/unit/financing-statement/financing-statment.spec.ts]: What about a term of 3.14159?

**Review by WalterMoar — CHANGES_REQUESTED** [review]
Some changes requested.

File: ppr-ui/tests/unit/financing-statement/financing-statment.spec.ts (hunk 12 above)
Comment (bryan-gilbert) [reply ppr-ui/tests/unit/financing-statement/financing-statment.spec.ts]: I thought the question to ask was to be or not to be but I prefer pie any day over existential pondering.

**Comment by bryan-gilbert** [comment]
thanks for the comments.  all adjustments made

File: ppr-ui/src/assets/styles/styles.scss [hunk 13]
```diff
@@ -14,3 +14,8 @@ and mixins there is no CSS code bloat.
     bottom: 0;
     width: 100%;
   }
+
+.app-body {
+  padding-bottom: 2rem;
+}
+
```
Comment (WalterMoar) [inline ppr-ui/src/assets/styles/styles.scss:21]: Hmm, missing newline snuck in.

File: ppr-ui/src/financing-statement/FinancingStatement.vue [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    function updateRegisteringParty(newPerson: PersonModel): void {
+      emit('input', new FinancingStatementModel(
+        props.value.type,
+        props.value.life,
+        newPerson // props.value.registeringParty
+      ))
+    }
+
+    // Callback function for emitting model changes made to the FS life
+    function updateTerm(newLife: number): void {
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/FinancingStatement.vue:96]: This life vs term thing should hopefully get cleared up by the UX work - good enough for now :)

File: ppr-ui/src/financing-statement/financing-statement-model.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+   * Creates a new FinancingStatementModel model instance.
+   *
+   * @param type the type of financing statement. A value from the FinancingStatementTypes enum
+   * @param life the number of years the financing statement is registered for. A value between 1 and 25.
+   * @param registeringParty the PersonModel who registered the financing statement
+   */
+  public constructor(
+    type = FinancingStatementTypes.SECURITY_AGREEMENT,
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/financing-statement-model.ts:19]: Hmm, are enums usually singular? Not a big deal just something to ponder.

File: ppr-ui/src/financing-statement/financing-statement-model.ts [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+  /**
+   * Gets the Person who registered the financing statement
+   */
+  public get registeringParty(): PersonModel {
+    return this._registeringParty
+  }
+
+  /*
+   * Class declarations
+   */
```
Comment (WalterMoar) [inline ppr-ui/src/financing-statement/financing-statement-model.ts:51]: Stealing this :)

File: ppr-ui/src/views/FinancingStatementView.vue [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+        </header>
+        <p>
+          <em>Note: This web site is a work in progress.</em>
+        </p>
+        <v-btn
+          text
+          small
+          @click="toggle()"
+        >
+          Toogle editing {{ editing }}
```
Comment (WalterMoar) [inline ppr-ui/src/views/FinancingStatementView.vue:21]: Toogle :)

**Review by WalterMoar — APPROVED** [review]
Niiice!

File: ppr-ui/src/assets/styles/styles.scss (hunk 13 above)
Comment (bryan-gilbert) [reply ppr-ui/src/assets/styles/styles.scss:21]: There is a new line it just happens that a space is added to that newline.

File: ppr-ui/src/financing-statement/FinancingStatement.vue (hunk 14 above)
Comment (bryan-gilbert) [reply ppr-ui/src/financing-statement/FinancingStatement.vue:96]: will fix now and will appear in next pr

File: ppr-ui/src/financing-statement/financing-statement-model.ts (hunk 15 above)
Comment (bryan-gilbert) [reply ppr-ui/src/financing-statement/financing-statement-model.ts:19]: will fix ... see next pr

**Pull request merged**